project-icp/bee-pollinator-app
# Scaffold Beekeeper Front-End

## Overview

Adds Ansible provisioning for Docker, which we use to build the Beekeepers front-end using Node 8, Webpack, and React. Use `django-webpack-loader` to be able to serve webpack made JavaScript files, including hashes and hot reloading, via Django. This allows us to use the optimization, translation, and hot reloading of webpack, while still having access to Django variables such as CSRF and use the same domain as the API, which makes API calls much simpler.

See commit messages for more details.

Tagging @mmcfarland and @hectcastro for provisioning review.

Connects #286 

### Demo

![image](https://user-images.githubusercontent.com/1430060/46681363-16ff0480-cbb9-11e8-8772-81ffed92d0a7.png)

### Notes

Is `beekeepers` the right keyword to use here (for the script, the Django app, the query string parameter)? Is it accurate? Is it too long?

There is a fair amount of indirection here: the `beekeepers.sh` script is triggered from the host, which goes in to Vagrant, which runs `yarn.sh`, which runs Docker, which runs `yarn`, which runs `webpack`. However, it works. If there is a simpler way of doing this, we should consider it.

## Testing Instructions

* Check out this branch
* Delete the `app` VM if it exists.
* Run `vagrant up` to create and provision the `app` VM, and start the others
* Once provisioning completes, run `./scripts/manage.sh migrate` to install migrations for the original Pollination Mapper app
* Go to [:8000/](http://localhost:8000/) and ensure you see the Pollination Mapper correctly
* Go to [:8000/?beekeepers](http://localhost:8000/?beekeepers) and ensure you see a sample React app
* Run `./scripts/beekeepers.sh run dev-server` and then go to [:8000/?beekeepers](http://localhost:8000/?beekeepers) again.
  - Open dev tools console, and ensure you see the redux log when you click the button.
  - Open [`SampleApp.jsx`](https://github.com/project-icp/bee-pollinator-app/blob/tt/scaffold-bees2-front-end/src/icp/apps/beekeepers/js/src/SampleApp.jsx) and edit some text, and save the file. Ensure you see it hot reload in the browser
* Run `./scripts/beekeepers.sh run bundle` to make the production bundle and ensure the app works again at [:8000/?beekeepers](http://localhost:8000/?beekeepers)
